artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQSession.java

```diff
@@ -361,6 +362,13 @@ public void run() {
             connection.getTransportConnection().setAutoRead(false);
          }
 
+         QueueQueryResult result = getCoreSession().executeQueueQuery(addresses[i]);
```

Comment:
We can change ServerSessionImpl::send as:

send(final ServerMessage message, final boolean direct, final boolean expectedRouting) throws Exception


we can down the road, change PostOfficeImpl::route to add expectedRouting parameter, and if the routeContext.getCount() == 0 and expectedRouting == true, then throw the exception.


Better than run this expensive query through the bindings on every message sent.